src/crystal/scheduler.cr

```diff
@@ -45,42 +56,178 @@ class Crystal::Scheduler
+
+    while Crystal.running?
+      # 1. yield CPU time to another thread
+      Thread.yield
+
+      # 2. try to steal a runnable fiber from a random scheduler:
+      fiber = steal_once
+
+      unless fiber
+        # 3. process the event loop to fill the runnables queue if any events
+        #    are pending, but don't block, since it would prevent this thread
+        #    from stealing fibers (or resuming a yielded fiber), and block other
+        #    threads, too:
+        Crystal::EventLoop.run_nonblock
+
+        # try to dequeue a fiber from the internal runnables queue:
+        fiber = @runnables.pop?
+      end
+
+      if fiber
+        # success: we now have a runnable fiber (dequeued, stolen or ready):
+        resume(fiber)
+
+        # the runnables queue is now empty (again)
+      end
+    end
+  end
+
+  protected def steal_once : Fiber?
+    runnables = @@all_runnables.sample(@rng)
+    runnables.shift? unless runnables == @runnables
   end
 
   protected def enqueue(fiber : Fiber) : Nil
-    @runnables << fiber
+    @runnables.push(fiber)
+    # TODO: wakeup a parked thread (if any)
   end
 
   protected def enqueue(fibers : Enumerable(Fiber)) : Nil
-    @runnables.concat fibers
+    fibers.each do |fiber|
+      @runnables.push(fiber)
+    end
+    # TODO: wakeup parked threads (if any)
   end
 
   protected def resume(fiber : Fiber) : Nil
+    # a fiber can be enqueued before saving its context, which means there is a
+    # time window for a scheduler to steal and try to resume a fiber before its
+    # context has been fully saved, so we wait:
+    spin_until_resumable(fiber)
+
+    # replace references:
     current, @current = @current, fiber
-    GC.stack_bottom = fiber.@stack_bottom
-    Fiber.swapcontext(pointerof(current.@stack_top), fiber.@stack_top)
+
+    {% unless flag?(:gc_none) %}
+      # register the fiber's stack as the current thread's stack to the garbage
+      # collector; we disable BDWGC to avoid a race-condition between
+      # register-gc-stack and swapcontext during which another thread could call
+      # GC_collect and stop-the-world, leaving the GC with an invalid thread
+      # stack:
+      GC.disable
+      GC.stack_bottom = fiber.@stack_bottom
+      # GC.set_stack_bottom(LibC.pthread_self, fiber.@stack_bottom)
+    {% end %}
+
+    # swap the execution context:
+    Fiber.swapcontext(pointerof(current.@context), pointerof(fiber.@context))
+
+    {% unless flag?(:gc_none) %}
+      # eventually resume the GC (in the new context)
+      GC.enable
+    {% end %}
+  end
+
+  private def spin_until_resumable(fiber)
+    # 1. 1st attempt will always succeed with a single thread (unless the fiber
+    #    is dead):
+    until fiber.resumable?
+      # 2. constant-time busy-loop, to avoid an expensive thread context switch:
+      #    relies on a magic number, but anything higher don't seem to improve
+      #    performance (on x86_64 at least):
+      99.times do
```

Comment:
Why trying  99 times very close to each other instead of several times(<10) with a wait time between?

Reply:
because he has 99 problems but a switch ain't one XD

Reply:
BTW: there are 100 attempts, not 99, one musn't forget the initial `until @flag.test_and_set`!

Spin locks must resolve as fast as possible, because they block the thread from progressing. I initially had a simple `until @flag.test_and; Thread.yield; end`. 

In a furious benchmark, with N threads running M fibers that each call `Fiber.yield` a fixed number of time, introducing a constant time busy-loop reduced thread contention, and thus spin lock overhead, by roughly 30% (if I remember correctly). Using 100 attempts is indeed a magic number, but used in different places (e.g. musl-libc pthread implementation), and using anything else, in my furious benchmark, didn't improve performance in a measurable manner.

Eventually, using an adaptive busy-loop will probably squeeze some more performance, but it will require real-world scenarios to benchmark correctly.

I recommend reading the [Empirical Studies of Competitive Spinning for A Shared-Memory Multiprocessor](https://www.researchgate.net/publication/234794598_Empirical_studies_of_competitve_spinning_for_a_shared-memory_multiprocessor) paper by Anna R. Karlin, Kai Li, Mark S. Manasse, Susan Owicki.

Reply:
Thanks for the reference! IIRC Fixed Spin is used here, and in the future Random walk could be considered?

Reply:
Maybe, in the future. With better benchmarks on real-world scenarios.

Random-walk looks really good in the paper, but the variable algorithm is much more complex to implement than the constant algorithm (requires timings, and probably some arch/os adjustments). In my experiments, I wasn't able to get the performance improvement (probably my mistakes), so I settled on the constant time busy loop, which was both simple & improved performance significantly.